File: src/main/java/ch/unibas/dmi/dbis/vrem/database/dao/VREMReader.java
package ch.unibas.dmi.dbis.vrem.database.dao;

import ch.unibas.dmi.dbis.vrem.database.codec.ArtCollectionCodec;
import ch.unibas.dmi.dbis.vrem.database.codec.ExhibitionCodec;
import ch.unibas.dmi.dbis.vrem.model.exhibition.Exhibit;
import ch.unibas.dmi.dbis.vrem.model.exhibition.Exhibition;
import ch.unibas.dmi.dbis.vrem.model.exhibition.ExhibitionSummary;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.Projections;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import org.bson.Document;
import org.bson.types.ObjectId;

public class VREMReader extends VREMDao {

    /**
     *
     */
    public VREMReader(MongoDatabase database) {
        super(database);
    }


    public Exhibition getExhibition(String name) {
        return getExhibition(ExhibitionCodec.FIELD_NAME_NAME, name);
    }

    public Exhibition getExhibition(ObjectId id) {
        return getExhibition(ExhibitionCodec.FIELD_NAME_ID, id);
    }

    private Exhibition getExhibition(String fieldName, Object key) {
        final MongoCollection<Exhibition> exhibitions = this.database.getCollection(EXHIBITION_COLLECTION, Exhibition.class);
        return exhibitions.find(Filters.eq(fieldName, key)).first();
    }

    /**
     *
     */
    public List<ExhibitionSummary> listExhibitions() {
        final MongoCollection<Document> exhibitions = database.getCollection(EXHIBITION_COLLECTION);
        final List<ExhibitionSummary> list = new ArrayList<>();
        for (Document document : exhibitions.find().projection(Projections.include(ExhibitionCodec.FIELD_NAME_ID, ExhibitionCodec.FIELD_NAME_NAME))) {
            list.add(new ExhibitionSummary(document.getObjectId(ExhibitionCodec.FIELD_NAME_ID), document.getString(ExhibitionCodec.FIELD_NAME_NAME)));
        }
        return list;
    }

    public List<Exhibit> listExhibits() {
        // Get all Exhibits from all ArtCollections.
        final MongoCollection<Document> artCollections = database.getCollection(CORPUS_COLLECTION);
        final List<Exhibit> list = new ArrayList<>(listExhibitsFromExhibitions());
        for (Exhibit e : artCollections.distinct(ArtCollectionCodec.FIELD_NAME_EXHIBITS, Exhibit.class)) {
            list.add(e);
        }
        return list.stream().distinct().collect(Collectors.toList());
    }

    public List<Exhibit> listExhibitsFromExhibitions(){
        final MongoCollection<Exhibition> exhibitions = database.getCollection(EXHIBITION_COLLECTION, Exhibition.class);
        final List<Exhibit> list = new ArrayList<>();
        exhibitions.find().forEach((Consumer<? super Exhibition>) e -> list.addAll(e.getExhibits()));
        return list;
    }
}

File: src/main/kotlin/ch/unibas/dmi/dbis/vrem/kotlin/database/dao/VREMReader.kt
package ch.unibas.dmi.dbis.vrem.kotlin.database.dao

import ch.unibas.dmi.dbis.vrem.kotlin.model.collection.ArtCollection
import ch.unibas.dmi.dbis.vrem.kotlin.model.exhibition.Exhibit
import ch.unibas.dmi.dbis.vrem.kotlin.model.exhibition.Exhibition
import ch.unibas.dmi.dbis.vrem.kotlin.model.exhibition.ExhibitionSummary
import com.mongodb.client.MongoDatabase
import org.bson.types.ObjectId
import org.litote.kmongo.*
import org.litote.kmongo.id.toId

/**
 * TODO: Write JavaDoc
 * @author loris.sauter
 */
class VREMReader(database: MongoDatabase) : VREMDao(database) {

    fun getExhibition(name: String): Exhibition {
        val col = getExhibitionCollection()
        return col.findOne(Exhibition::name eq name)!!
    }

    fun getExhibition(id: ObjectId): Exhibition {
        val col = getExhibitionCollection()
        return col.findOne { Exhibition::id eq id.toId() }!!
    }


    fun listExhibitions(): List<ExhibitionSummary> {
        val col = getExhibitionCollection()
        return col.find().projection(Exhibition::id, Exhibition::name).map { ExhibitionSummary(it.id.toString(), it.name) }.toMutableList()
    }

    fun listExhibits(): List<Exhibit> {
        val artCollections = database.getCollection<ArtCollection>(CORPUS_COLLECTION)
        return artCollections.distinct(ArtCollection::exhibits).first()!!
    }

    fun listExhibitsFromExhibitions(): List<Exhibition> {
        val exhibitions = database.getCollection(EXHIBITION_COLLECTION, Exhibition::class.java)
        return exhibitions.find().toMutableList()
    }
}